### ISSUE TITLE
No support for RNFR and RNTO commands to rename or move files

### ISSUE THREAD

**Issue by fiasko131** [hxcan/ftpserver#16]
Here is what I suggest that works for me:

add in the  processCommand method of ControlConnectHandler class:

````java
public void processCommand(String command, String content, boolean hasFolloingCommand) {
	......
    else if (command.equalsIgnoreCase("RNFR")){
        String data51= content.substring(5);
        data51=data51.trim();
        processRenameFromCommand(data51);
    }
    else if (command.equalsIgnoreCase("RNTO")){
        String data51= content.substring(5);
        data51=data51.trim();
        processRenameToCommand(data51);
    }
	..........
}
````

add methods:
````java
/**
   *  Process the rename command
   */
    private void processRenameFromCommand(String data51){
      if (!allowWrite){ // A boolean that I added in the case where writing to the server is not authorized
        binaryStringSender.sendStringInBinaryMode("550 Writing permission not granted.");
        return;
      }
      Log.d("renameCommandFrom","data51 "+data51);
      Log.d("renameCommandFrom","rootDirectory "+rootDirectory);
      Log.d("renameCommandFrom","currentWorking "+currentWorkingDirectory);
      currentWorkingDirectory = currentWorkingDirectory.replace("//","/");
      // add a field oldPath
      if (data51.startsWith("/")){
        this.oldPath = rootDirectory.getPath() +data51;
      }else {
        this.oldPath = rootDirectory.getPath() + currentWorkingDirectory+"/"+data51;
      }
      Log.d("renameCommand","oldPath "+oldPath);
      binaryStringSender.sendStringInBinaryMode("350 File to rename: "+currentWorkingDirectory+"/"+data51);
    }

  /**
   *  Process the rename command
   */
  private void processRenameToCommand(String data51){
    Log.d("renameCommandTo","data51 "+data51);
    String newPath = rootDirectory.getPath()+data51;
    Log.d("renameCommandTo","newPath "+newPath);
    if (renameFile(this.oldPath,newPath)){
      binaryStringSender.sendStringInBinaryMode("250 Success! renamed to: "+data51);
    }else {
      binaryStringSender.sendStringInBinaryMode("500 Failed! renamed to: "+data51);
    }

  }

  public static boolean renameFile(String oldFileName, String newFileName) {
    File oldFile = new File(oldFileName);
    File newFile = new File(newFileName);

    if (oldFile.exists()) {
      boolean success = oldFile.renameTo(newFile);
      return success;
    } else {
      return false; // Le fichier source n'existe pas.
    }
  }
````
allowWrite is a boolean that I added to the FtpServer and ConneConnectHandler constructor for my implementation.

Thanks!

**Comment by hxcan** [comment]
I have to choose a license for this project and then you can create a pull request for this.

Also, I need to read the documents about the meaning of these two commands as I don't know every details about the ftp standard.

**Comment by hxcan** [comment]
I have chosen the MIT license.

Is it OK for you to commit your changes as a pull request?

**Comment by fiasko131** [comment]
It's just a suggestion, I added these methods in my implementation, and they work perfectly for me. RNFR and RNTO are standardized FTP protocol commands:
https://www.serv-u.com/resources/tutorial/rmda-xrmd-rest-rnfr-rnto-thmb-ftp-command

What do you mean by: "I have to choose a license for this project"?

**Comment by fiasko131** [comment]
> I have chosen the MIT license.

I don't know how to use github very well and I've never made a pull request... :((

**Comment by hxcan** [comment]
I'm not a lawyer.
But I have heard that:
Without an explicit license, the copyright law applies here. You might be in trouble when using my project, and I might be in trouble when using the code in your pull request, which will prevent me from accpeting your pull request at all.

https://opensource.stackexchange.com/a/1721

**Comment by hxcan** [comment]
> > I have chosen the MIT license.
> 
> I don't know how to use github very well and I've never made a pull request... :((

That's OK. I can implement the commands when I have finished learning them.

**Comment by fiasko131** [comment]
But your presentation page does not indicate any license?

**Comment by hxcan** [comment]
Reload the index page:
https://github.com/hxcan/ftpserver

You will see a "MIT License" line in the "About".
You can also see a "LICENSE" file in the file list.

**Comment by fiasko131** [comment]
I'm not a specialist but when I downloaded the files from your library there was no license indicated. So how to do it? Do I need to ask you for a license?

**Comment by hxcan** [comment]
Pull(or download) the master branch from my project again and do your work based on that revision, specificly this revision 1078ad9: https://github.com/hxcan/ftpserver/commit/1078ad976b7efd5223f530bfef29569ee868941b

Any revision after that commit will have the license file inside, and I believe no further "license work" will be required.

**Comment by fiasko131** [comment]
Ok, I hope that adding the methods for moving files will be operational for your library

Question: Why do you work with URIs since your library can only work with an application which must require access to all files:
``
before android 10: "<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />"
android 10: "<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />" + android:requestLegacyExternalStorage="true"
android 11 and above: "<uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />"
``

**Comment by hxcan** [comment]
We support some uri style resources on Android:
https://github.com/hxcan/ftpserver/issues/10
This is not limited to actual files, in theory.

Even for the actual files, some directories can only be accessed through a browse action by the user, for example: EXTERNAL_STORAGE/Android/data
The directory tree returned from a "browse" action on Android, will be an Android-only uri.

And I have this feature in my wish list:
The HxFtpServer will accept files shared from other Android applications, and make them downloadable via some FTP urls. Those files shared from other applications will be uri style resources, too.

**Comment by hxcan** [comment]
We support some uri style resources on Android:
https://github.com/hxcan/ftpserver/issues/10
This is not limited to actual files, in theory.

Even for the actual files, some directories can only be accessed through a browse action by the user, for example: EXTERNAL_STORAGE/Android/data
The directory tree returned from a "browse" action on Android, will be an Android-only uri.

And I have this feature in my wish list:
The HxFtpServer will accept files shared from other Android applications, and make them downloadable via some FTP urls. Those files shared from other applications will be uri style resources, too.

**Comment by hxcan** [comment]
In place renaming is supported for a target file name in the same parent path.

It is implemented in the latest version "2023.10.32".
